### PR TITLE
new hw_proc_id plugin and fixed up r28 logic

### LINKED ISSUE #990: MIPS OSI doesn't autoload correctly from pypanda
For architectures other than mips, this script will print a non-null proc object regardless if the `bug` flag is set. However, for mips, it will only print a valid value if `bug` is false.

```py
from pandare import Panda

bug = True
arch = "mips"
panda = Panda(generic=arch)

@panda.queue_blocking
def driver():
    panda.revert_sync("root")

    if not bug:
        panda.load_plugin("osi")

    @panda.cb_before_block_exec
    def bbe(cpu, tb):
        proc = panda.plugins['osi'].get_current_process(cpu)
        print(proc)
        if proc != panda.ffi.NULL:
            panda.disable_callback("bbe")

panda.run()
```

### PR DESCRIPTION
**Pull request by lacraig2**
- Adds new plugin `hw_proc_id` from @AndrewFasano
- changes first_osi_check switch to syscall method to wait until hw_proc_id has been initialized on MIPS
- moves MIPS r28 cache in osi_linux to `hw_proc_id` and changes cache callback to bbe to sbe
- adds new `address_in_kernel_code_linux` from functionality in `panda_in_kernel_code_linux`
- reorganizes r28 cache code in `hw_proc_id` to use new call (address now supports MIPS64)

CLOSES: #990

### CODE REVIEW

**Comment by AndrewFasano** [comment]
Any chance you could just pull in #1142 with this? They seem to overlap a bit

**Comment by AndrewFasano** [comment]
Sorry, I haven't had time to address the changes requested there, but it seems like you've made some of them here

**Comment by lacraig2** [comment]
Sure. I've pulled in that one too.

**Comment by AndrewFasano** [comment]
Other than the CI failures, this looks great to me. Thanks so much for pulling in the code from that other PR. I swear I was going to address your feedback soon, but I'm having a bit of a busy month right now.

File: panda/plugins/hw_proc_id/hw_proc_id_int_fns.h [hunk 1]
```diff
@@ -0,0 +1,13 @@
+#pragma once
+
+// BEGIN_PYPANDA_NEEDS_THIS -- do not delete this comment bc pypanda
+// api autogen needs it.  And don't put any compiler directives
+// between this and END_PYPANDA_NEEDS_THIS except includes of other
+// files in this directory that contain subsections like this one.
+
+unsigned int get_id(CPUState *cpu);
+
+bool is_is_initialized(void);
```
Comment (off-by-1-error) [inline panda/plugins/hw_proc_id/hw_proc_id_int_fns.h]: id_is_initialized?

**Review by off-by-1-error — CHANGES_REQUESTED** [review]
small typo in `hw_proc_id_int_fns.h` 

everything else looks good!

File: panda/plugins/hw_proc_id/hw_proc_id_int_fns.h (hunk 1 above)
Comment (lacraig2) [reply panda/plugins/hw_proc_id/hw_proc_id_int_fns.h]: Good catch. Thanks.

File: panda/plugins/osi_linux/osi_linux.cpp [hunk 2]
```diff
@@ -285,6 +302,16 @@ bool osi_guest_is_ready(CPUState *cpu, void** ret) {
     // If it's the very first time, try reading current, if we can't
     // wait until first sycall and try again
     if (first_osi_check) {
+        #ifdef TARGET_MIPS
+        if (!get_id(cpu)){
+            // If we're on MIPS, we need to wait until r28 is set before
+            // moving to a syscall strategy
+            if (!id_is_initialized()){
+                ret = NULL;
```
Comment (MarkMankins) [inline panda/plugins/osi_linux/osi_linux.cpp:310]: Should this (and line 336 below) be:  *ret = NULL;    ?

**Pull request merged**